File: January2022/Java/RobotBoundedInCircle.java
class Solution {
	// TC: O(len(instructioons))
	// SC: O(1)
	private enum Direction {
		NORTH,
		SOUTH,
		WEST,
		EAST
	}

	public boolean isRobotBounded(String instructions) {
		int x = 0, y = 0;// starting
		Direction direction = Direction.NORTH;

		for (char instruction : instructions.toCharArray()) {
			if (instruction == 'G') {
				switch (direction) {
					case NORTH:
						y++;
						break;
					case SOUTH:
						y--;
						break;
					case WEST:
						x--;
						break;
					case EAST:
						x++;
						break;
				}
			} else if (instruction == 'L') {
				switch (direction) {
					case NORTH:
						direction = Direction.WEST;
						break;
					case SOUTH:
						direction = Direction.EAST;
						break;
					case WEST:
						direction = Direction.SOUTH;
						break;
					case EAST:
						direction = Direction.NORTH;
						break;
				}
			} else if (instruction == 'R') {
				switch (direction) {
					case NORTH:
						direction = Direction.EAST;
						break;
					case SOUTH:
						direction = Direction.WEST;
						break;
					case WEST:
						direction = Direction.NORTH;
						break;
					case EAST:
						direction = Direction.SOUTH;
						break;
				}
			}
		}

		if (x == 0 && y == 0) { return true; }
		if (direction == Direction.NORTH) { return false; }

		return true;
	}

}

File: January2022/Python/RobotBoundedInCircle.py
#Author :  Karthik K ( https://www.linkedin.com/in/karthikmunirathinam/ )

# TC : O(n)
# SC : O(1)

class Solution:
    def isRobotBounded(self, instructions: str) -> bool:
        x,y=0,0
        dirX,dirY=0,1
        for direction in instructions:
            if direction=='G':
                x,y=x+dirX,y+dirY
            elif direction=='L':
                dirX,dirY=-1*dirY,dirX
            else:
                dirX,dirY=dirY,-1*dirX
        return (x,y)==(0,0) or (dirX,dirY)!=(0,1)